README.md
# mlp-emoticons-telegram
Emoticons copy from reddit https://www.reddit.com/r/mylittlepony/ .
Power by [telegraf.js](https://github.com/telegraf/telegraf).

### Features
1.Search results by name. (if you want fuzzy search, please do you own. )

2.Get random results if you query a word not in list(or database,something else, etc...)

### make your own
If you want to get your own bot. Please read Suggestions below.

1.You can generator the path json through the [path2json.js](https://github.com/instkffff/mlp-emoticons-telegram/blob/master/path2json.js), this will be use to query picture in your file-server.

2.Write or generator a index list for the json file you generator in first step. You'll be find specification in index.js .

3.You can hosting your picture use [path2json.js](https://github.com/zeit/serve-handler). I have gave you an example [server.js](https://github.com/instkffff/mlp-emoticons-telegram/blob/master/emotes/100x100/serve.js)

4.replace the picture url by your own, or you can rewrite the index.js in order to configuration it in [conf.env](https://github.com/instkffff/mlp-emoticons-telegram/blob/master/conf.env). You can delete the 'photo_width','photo_height' set for different images size. You also can modifier by your own though read the telegram official doc https://core.telegram.org/bots/api#answerinlinequery.

5.replace the bot token in [conf.env](https://github.com/instkffff/mlp-emoticons-telegram/blob/master/conf.env), and get it start in your own way , have fun!


rename tool/History.txt
ReNamer 6.9 (2018-06-30)
~~~~~~~~~~~~~~~~~~~~~~~~

6.8.0.5 Beta (2018-06-27)
* Fixed (again) incorrectly sized tool buttons in the Pascal Script window. Issue introduced in v6.8.0.1 Beta.

6.8.0.4 Beta (2018-06-24)
* Implemented rule comments, editable via rules context menu.
* Enhanced storage path resolution with new path tags and handling of multiple path tags.
* Added more unit tests for rules.
* Updated Russian language file.

6.8.0.3 Beta (2018-06-10)
* Major code refactoring to improve automation and unit testing capabilities.
* Decoupled renaming rules from the visual components.
* Refactored renaming rule structures and management operations.
* Refactored file signature structures and management operations.
* Fixed incorrectly sized tool buttons in the Pascal Script window. Issue introduced in v6.8.0.1 Beta.
* Added a dialog for manually entering file paths for renaming.
* Added various unit tests.
* Added Traditional Chinese (Taiwan) language file. Thanks to 冰霧神域.
* Updated German language file. Thanks to Kevin Scheitler.
* Updated Serbian language file. Thanks to Никола Алексић.
* Updated Russian language file.

6.8.0.2 Beta (2018-01-12)
* Remove all files more efficiently than one-by-one when all files match the clearing criteria.
* Efficiency improvements for files table clearing operations.
* Efficiency improvements for performing shell actions on files.
* Internal refactoring.

6.8.0.1 Beta (2018-01-08)
* Added "How to translate" link to the Languages menu.
* Find application language file independently of the executable file name.
* Perform a fast clearing of files tables when closing the application.
* Converted action buttons in the Pascal Script window to a toolbar. Better handling of long localized text.
* Changed the Z-order of controls in the Delete rule window. Avoid awkward overlapping with long localized text.
* Forbid the use of empty folder when creating preset links. Avoid an error.
* Added Dutch language file.
* Updated French language file.
* Updated Chinese Simplified language file. Thanks to WanderMax.


ReNamer 6.8 (2017-11-18)
~~~~~~~~~~~~~~~~~~~~~~~~

6.7.0.9 Beta (2017-11-15)
* Optimized files table clearing operation. Clearing all files is significantly faster now. 
* Refactored files table import operation.

6.7.0.8 Beta (2017-11-05)
* Include directories with a system attribute when scanning for presets.
* Show presets directory structure limitation message in Save Preset and Presets Manager.
* Reapply presets limitations after a preset was deleted in Presets Manager.
* Localized filter title in the dialog for importing presets.
* Updated Russian language file.

6.7.0.7 Beta (2017-10-12)
* Ignore file extension when sorting presets.
* Sort the list of translit alphabets.
* Meta tags: Decoded escape sequences when extracting complete mailbox specifications.
* Meta tags: Handle tab char (#9) as a fold char in email headers.
* Updated Italian language file. Thanks to Vittorio Ierardi.

6.7.0.6 Beta (2017-09-20)
* Meta tags: Handle folded (wrapped/multiline) email headers.
* Meta tags: Handle escape sequences in mailbox specification.
* Meta tags: Decode multiple contained MIME encoded words in email headers.

6.7.0.5 Beta (2017-09-14)
* Meta tags: Decode MIME encoded email subject line and addressee names.
* Meta tags: Extract unquoted name from a mailbox addressee specification.
* A more user-friendly error message when the application of renaming rules on text fails.
* Rearranged Z-order of options to avoid covering up of the radio button in the User Input rule, due to increased system font size.
* Changed the order of options in the Serialize rule for a more natural flow.
* Display separate counters for skipped unmarked files and skipped unchanged files.
* Updated Russian language file.

6.7.0.4 Beta (2017-06-28)
* Added new functions to Pascal Script: WideMatchesMask, WideMatchesMaskList, WideSameFileName, WideReverseString, RandomString.
* Refactored registration of custom types and functions in Pascal Script.
* Other internal changes.

6.7.0.3 Beta (2017-06-10)
* Support directory structures for presets (Pro version).
* Do not show "preset not found" warning in silent command line mode.
* Apply license restrictions to the creation of preset links.
* Show message with the number of successfully created preset links.
* Save Preset dialog is now resizable.
* Allow ampersand sign "&" to be used in preset names.
* Forbid saving a preset with an empty name.
* Updated Russian language file.

6.7.0.2 Beta (2017-05-31)
* Added a selection of numeral systems to the Serialize rule: Decimal digits (0..9), English letters (a..z), Roman numerals (I,II,III,IV,...), Simplified Chinese, Custom alphabetic and Custom numeric.
* Avoid a crash when Randomize rule is misconfigured with an empty character set.
* Refactored UTF8 code point splitting routines.
* Updated Russian language file.

6.7.0.1 Beta (2017-05-26)
* Added "Reset if file name changes" option to the Serialize rule.
* Added FileReadTextLines function to Pascal Script.
* Added UsePlusAsSpace parameter to URLDecode function.
* Modified import/export icons to minimize ambiguity of color and direction of arrows.
* Added "New Instance" option for opening a new application instance.
* Option "Create Links" can now handle Unicode preset names.
* Option to add a link to "Send To" folder is now Unicode capable.
* Updated Japanese language file. Thanks to Tilt.
* Updated Chinese Simplified language file. Thanks to Max Wander.


ReNamer 6.7 (2017-03-29)
~~~~~~~~~~~~~~~~~~~~~~~~

6.6.0.7 Beta (2017-02-23)
* Strip surrounding white space characters from EXIF Make, Model and Description tags.
* Display fractional part in "Size KB" and "Size MB" columns in the files table.
* Format file size and torrent size meta tags using locale symbols and 2 fractional digits.
* Added "Torrent_TotalSize", "File_Size" and "File_SizeGB" meta tags.
* Improved routines for number formatting based on locale settings.

test/query.js
const pony = require('../pony.json')

var list = {
	'applebloom' : 0 ,
	'applejack' : 1 ,
	'cadance': 2,
	'celestia': 3,
	'derpy': 4,
	'discord': 5,
	'fluttershy': 6,
	'luna': 7,
	'other': 8,
	'pinkie': 9,
	'rainbowdash': 10,
	'rarity': 11,
	'scootaloo': 12,
	'spike': 14,
	'starlight': 15,
	'students': 16,
	'sunset': 17,
	'sweetiebelle': 18,
	'tempest': 19,
	'trixie': 20,
	'twilight': 21,
	'yark': 22
}

Pony = 'twilight'

console.log(typeof(Pony))

Ponyid = list[Pony]

console.log(Ponyid)



console.log(list.hasOwnProperty(Pony))

Pony_id = parseInt(Ponyid)



ponys = pony[Pony_id].children


var item = Object.keys(list)[Math.floor(Math.random()*Object.keys(list).length)];



console.log(parseInt(list[item]))



console.log(Object.keys(ponys).length)

test/inline.js
const Telegraf = require('telegraf')
const fetch = require('node-fetch')
require('dotenv').config({path:'../conf.env'})

// async/await example.

async function omdbSearch (query = '') {
  const apiUrl = `http://www.omdbapi.com/?s=${query}&apikey=9699cca`
  const response = await fetch(apiUrl)
  const json = await response.json()
  const posters = (json.Search && json.Search) || []
  return posters.filter(({ Poster }) => Poster && Poster.startsWith('https://')) || []
}

const bot = new Telegraf(process.env.BOT_TOKEN)

bot.on('inline_query', async ({ inlineQuery, answerInlineQuery }) => {
  const posters = await omdbSearch(inlineQuery.query)
  const results = posters.map((poster) => ({
    type: 'photo',
    id: poster.imdbID,
    caption: poster.Title,
    description: poster.Title,
    thumb_url: poster.Poster,
    photo_url: poster.Poster
  }))
  return answerInlineQuery(results)
})

bot.launch()

path2json.js
  var fs = require('fs');
  var path = require('path');

  function dirTree(filepath) {
    var info = { name: path.basename(filepath) };
    var invalidFile = (info.name === '.DS_Store');
    var isThumbnail = (info.name.indexOf('_thumbnail') > -1 ||
                       info.name.indexOf('_sm') > -1);

    if (invalidFile || isThumbnail) {
      return; // this leaves an empty array element that we'll clean up on recursion
    }

    info.slug = info.name.toString().toLowerCase()
                    .replace(/\.[^/.]+$/, '') // Strip file extension
                    .replace(/\s+/g, '-')     // Replace spaces with -
                    .replace(/[^\w\-]+/g, '') // Remove all non-word chars
                    .replace(/\-\-+/g, '-')   // Replace multiple - with single -
                    .replace(/^-+/, '')       // Trim - from start of text
                    .replace(/-+$/, '');      // Trim - from end of text

    if (fs.lstatSync(filepath).isDirectory()) {
      info.path = filepath.replace('public', '');

      info.children = fs.readdirSync(filepath).map(function(child) {
        return dirTree(filepath + '/' + child);
      });

      // removes empty array elements resulting from early return above
      info.children = info.children.filter(function(e) { return e; });
    } else {
      // thumbnails should exist as a property of the gallery image itself, rather than adjacent neighbors in the collection
      // info.thumb = info.name.replace(/\.[^/.]+$/, '') + '_thumbnail.' + info.name.split('.').pop();
    }

    return info;
  }

  var gallery = dirTree('emotes/100x100');
  fs.writeFileSync('pony.json', JSON.stringify(gallery.children));


emotes/png2gif.py
#!/usr/bin/env python

# Convert PNG images to GIF, preserving transparency
# 2008 - http://www.coderholic.com/png2gif

import sys
from PIL import Image 
import random
import optparse
import os
import os.path

def unique_color(image):
	"""find a color that doesn't exist in the image
	"""
	colors = image.getdata()
	while True:
		# Generate a random color
		if image.mode == "LA":    
			 color = random.randint(0, 255),
		else:
			color = (
			  random.randint(0, 255),
			  random.randint(0, 255),
			  random.randint(0, 255)
			)

		if color not in colors:
			return color

def fill_transparent(image, color, threshold=0): 
	"""Fill transparent image parts with the specified color 
	"""
	def quantize_and_invert(alpha):
		if alpha <= threshold:
			return 255
		return 0
	# Get the alpha band from the image
	if image.mode == 'RGBA':
		red, green, blue, alpha = image.split()
	elif image.mode == 'LA':
		gray, alpha = image.split()
	# Set all pixel values below the given threshold to 255,
	# and the rest to 0
	alpha = Image.eval(alpha, quantize_and_invert)
	# Paste the color into the image using alpha as a mask
	image.paste(color, alpha)


def color_index(image, color):
	"""Find the color index
	"""
	palette = image.getpalette()
	palette_colors = zip(palette[::3], palette[1::3], palette[2::3])
	index = palette_colors.index(color)
	return index

def convert_image(image, name, new_name):
	if image.mode == 'P':
		if image.info.has_key('transparency'): # check to see if the image has any transparency
			transparency = image.info['transparency']
			image.save(new_name, transparency=transparency)
		else: image.save(new_name)
	elif image.mode == 'RGBA': # RGB images need to be converted to Palette mode
		threshold = 0
		colour = unique_color(image)
		fill_transparent(image, colour, threshold)
		image = image.convert('RGB').convert('P', palette=Image.ADAPTIVE)
		image.save(new_name, transparency=color_index(image, colour))
	elif image.mode == 'LA':
		threshold = 0
		colour = unique_color(image)
		fill_transparent(image, colour, threshold)
		image = image.convert('L').convert('P', palette=Image.ADAPTIVE)
		image.save(new_name, transparency=color_index(image, (colour[0], colour[0], colour[0])))
	else:
		raise "Unsupported PNG file" + name + "(" + image.mode + ")"

p = optparse.OptionParser(description = 'Convert PNG images to GIF format',
						  usage = 'png2gif [OPTIONS] <files>')
p.add_option('--outputdir', '-o', default='.', help='Set the output directory in which to put the GIF images. Defaults to the current directory')
p.add_option('--threshold', '-t', default='0', help='Set the transparency threshold. Defaults to 0')
p.add_option('--replace', '-r', action='store_true', help='Delete the PNG files after converting them to GIF')
p.add_option('--verbose', '-v', action='store_true', help='Verbose output')

options, arguments = p.parse_args()

# Check that the specified output directory exists
if not os.path.exists(options.outputdir):
	if options.versbose: print "Creating output directory", options.outputdir
	try:
		os.mkdir(options.outputdir)
	except:
		print "Cannot create output directory", options.outputdir
		print "Please specifiy a different directory"
		exit

for x in arguments:
	# Make sure we have the correct extension
	extension = x[-3:]
	if extension != "png":
		print "Invalid file extension for ", x, ". Expecting png found ", extension
		continue
	image = Image.open(x)
	if options.verbose:
		print "Converting", x, "to", x[:-3] + "gif"
	try:	
		convert_image(image, x, options.outputdir +os.sep + x[:-3] + "gif")
		if options.replace:
			if options.verbose: print "Removing", x
			os.remove(x)
	except:
		print "Problem converting",x, "- skipping"


index.js
const Telegraf = require('telegraf')
require('dotenv').config({path:'conf.env'})

const bot = new Telegraf(process.env.BOT_TOKEN)

const pony = require('./pony.json')

bot.on('inline_query', async({ inlineQuery, answerInlineQuery}) => {
  
  var list = {
  'applebloom' : 0 ,
  'applejack' : 1 ,
  'cadance': 2,
  'celestia': 3,
  'derpy': 4,
  'discord': 5,
  'fluttershy': 6,
  'luna': 7,
  'other': 8,
  'pinkie': 9,
  'rainbowdash': 10,
  'rarity': 11,
  'scootaloo': 12,
  'spike': 14,
  'starlight': 15,
  'students': 16,
  'sunset': 17,
  'sweetiebelle': 18,
  'tempest': 19,
  'trixie': 20,
  'twilight': 21,
  'yark': 22
  }

  const offset = parseInt(inlineQuery.offset) || 0 

  const results = []

  if(inlineQuery.query.length === 0){
    Folder = Object.keys(list)[Math.floor(Math.random()*Object.keys(list).length)]
    Pony_id = parseInt(list[Folder])
  }else if(list.hasOwnProperty(inlineQuery.query)){
    Pony_id = parseInt(list[inlineQuery.query])
    Folder = inlineQuery.query
  }else{
    Folder = Object.keys(list)[Math.floor(Math.random()*Object.keys(list).length)]
    Pony_id = parseInt(list[Folder])
  }
 
  
  var ponys = pony[Pony_id].children

  var num = Object.keys(ponys).length;

  for ( i = 0; i < num; i++ ){
    var Ponys = ponys[i]
    var url = `https://200x200-bibtvvfugm.now.sh/${Folder}/${Ponys.name}`
    results.push({
      type: 'photo',
      id: Ponys.slug,
      thumb_url: url,
      photo_url: url,
      photo_width: 200,
      photo_height: 200
    })  
  } 

  return answerInlineQuery(results, { next_offset: offset + 10 } , { cache_time: 10 })

})

bot.launch()
